modules/highgui/src/window_framebuffer.cpp

```diff
@@ -0,0 +1,800 @@
+{
+    return std::make_shared<FramebufferBackend>();
+}
+
+static std::string& getFBMode()
+{
+    static std::string fbModeOpenCV =
+    cv::utils::getConfigurationParameterString("OPENCV_HIGHGUI_FB_MODE", "FB");
+    return fbModeOpenCV;
+}
+
+static std::string& getFBFileName()
+{
+    static std::string fbFileNameFB =
+    cv::utils::getConfigurationParameterString("FRAMEBUFFER", "/dev/fb0");
+    static std::string fbFileNameOpenCV =
+    cv::utils::getConfigurationParameterString("OPENCV_HIGHGUI_FB_DEVICE", "");
+
+    if (!fbFileNameOpenCV.empty()) return fbFileNameOpenCV;
+    return fbFileNameFB;
+}
+
+FramebufferWindow::FramebufferWindow(FramebufferBackend &_backend, int _flags):
+    backend(_backend), flags(_flags)
+{
+    CV_LOG_DEBUG(NULL, "UI: FramebufferWindow::FramebufferWindow()");
+    FB_ID = "FramebufferWindow";
+    windowRect = Rect(0,0, backend.getFBWidth(), backend.getFBHeight());
+}
+
+FramebufferWindow::~FramebufferWindow()
+{
+    CV_LOG_DEBUG(NULL, "UI: FramebufferWindow::~FramebufferWindow()");
+}
+
+void FramebufferWindow::imshow(InputArray image)
+{
+    CV_LOG_DEBUG(NULL, "UI: FramebufferWindow::imshow(InputArray image)");
+    currentImg = image.getMat().clone();
+
+    CV_LOG_INFO(NULL, "UI: InputArray image: "
+    << cv::typeToString(image.type()) << " size " << image.size());
+
+    if (currentImg.empty())
+    {
+        CV_LOG_WARNING(NULL, "UI: image is empty");
+        return;
+    }
+    CV_CheckEQ(currentImg.dims, 2, "UI: dims != 2");
+
+    Mat img = image.getMat();
+    switch (img.channels())
+    {
+        case 1:
+            {
+                Mat tmp;
+                switch(img.type())
+                {
+                    case CV_8U:
+                        tmp = img;
+                        break;
+                    case CV_8S:
+                        cv::convertScaleAbs(img, tmp, 1, 127);
+                        break;
+                    case CV_16S:
+                        cv::convertScaleAbs(img, tmp, 1/255., 127);
+                        break;
+                    case CV_16U:
+                        cv::convertScaleAbs(img, tmp, 1/255.);
+                        break;
+                    case CV_32F:
+                    case CV_64F: // assuming image has values in range [0, 1)
+                        img.convertTo(tmp, CV_8U, 255., 0.);
+                        break;
+                }
+                Mat rgb(img.rows, img.cols, CV_8UC3);
+                cvtColor(tmp, rgb, COLOR_GRAY2RGB);
+                img = rgb;
+            }
+            break;
+        case 3:
+        case 4:
+            {
+                Mat tmp(img.rows, img.cols, CV_8UC3);
+                convertToShow(img, tmp, true);
+                img = tmp;
+            }
+            break;
+        default:
+            CV_LOG_ERROR(NULL, "UI: imshow can't convert image to show");
+            CV_Assert(img.channels() < 1);
+            CV_Assert(img.channels() > 4);
```

Comment:
Perhaps these asserts are ambiguous, we already know that image has wrong number of channels, they can be replaced with a single `CV_Error`:
```suggestion
            CV_Error(cv::Error::StsBadArg, "Bad image: wrong number of channels");
```